### DISCUSSION TITLE
custom plugin - set an upstream as a http fallback server

### DISCUSSION THREAD

**Discussion by ryan4yin** [apache/apisix#7773] [Q&A] [answered]
I want to write a custom plugin in lua, to configure an upstream as a http fallback server, the logic is:
-  if the default upstream's response status_code != 2xx, then request the fallback server, and forward the fallback server's response back to the client.

I looked at the builtin plugin `response-rewrite`, `batch-requests` and `proxy_mirror`, but still not sure how to implement it.

beacuse I need to check the default upstream's response status_code, I think I should execute my fallback-plugin at `_M.header_filter` and `_M.body_filter`, the order maybe like this:

- `_M.header_filter`: 
  - check the response code is not 2xx
  - request the fallback server using the client's arguments
  - set the`ngx.header` and `ngx.status` to the fallback server's response headers & status_code
  - store the fallback server's response body into ctx.var
- `_M.body_filter`
  -  set the`ngx.arg[1]` to the fallback server's response headers & status_code

any suggestions? thanks~

**Comment by tzssangglass** [comment, marked as answer]
we have support this feature, like

```
Each node can be configured with a priority. A node with low priority will only be used when all the nodes with higher priority have been tried or are unavailable.

As the default priority is 0, nodes with negative priority can be configured as a backup.
```

ref: https://github.com/apache/apisix/blob/master/docs/en/latest/admin-api.md#request-body-parameters-3

**Reply by ryan4yin** [reply]
Thanks for reply.
I would like to ask under what circumstances will the request fall back to the backup nodes with priority=-1, is it customizable?

I need to make sure that all non-2xx responses triggle a  fallback back operation to the backup server, because our default server is a new service in alpha stage, it's not reliable, and may make  mistakes which caused non-2xx status code.

**Reply by tzssangglass** [reply]
> I would like to ask under what circumstances will the request fall back to the backup nodes with priority=-1, is it customizable?

when all high-priority nodes are unavailable.

this is not customizable.

> I need to make sure that all non-2xx responses triggle a fallback back operation to the backup server, because our default server is a new service in alpha stage, it's not reliable, and may make mistakes which caused non-2xx status code.

I don't think there is any existing feature that can do this, it looks a lot like retry but is more advanced than what retry does.

Maybe you can try combining `retry` and `priority=-1` first.

**Reply by ryan4yin** [reply]
Yes, that's why I planned to implement a custom plugin to do this job, and then ask this question here. 
retry and negative priority seems not well documented, I do not think it's a good idea.

**Reply by tzssangglass** [reply]
Just a thought: the retry is the Nginx retry mechanism, and the plugin cannot dive into the Nginx retry mechanism.

about Nginx retry: https://developpaper.com/the-power-of-http-protocol-in-nginx-failure-retry-non_idempotent/

**Reply by ryan4yin** [reply]
thanks, it works~

I resolve this by add `proxy_next_upstream error timeout invalid_header http_500 http_502 http_503 http_504 http_429 non_idempotent;` into APISIX config's `configurationSnippet.httpSrv`, and then set my backup server's priority to `-1`, every thing works fine~

But I have another problem now: APISIX's prometheus plguin seems do not have any metrics about retry, I can not monitor how frequently the retries happens through metrics.
It's not a big problem, because the access logs can records all the `response_status` in order, I just curious is it possible to records the retries metrics in prometheus plugin.

**Reply by tzssangglass** [reply]
Sounds like it makes sense, let's see if more people will need it, you can also create issues to describe this feature.

BTW, I think you've made clever use of retry and backup servers.

**Reply by ryan4yin** [reply]
in case anyone need this, I adjust the `proxy-mirror`'s` plugin-attr`, set all the timeout too `300ms` to avoid the original requests wait for the mirrored request to complete.

**Reply by tokers** [reply]
> thanks, it works~
> 
> I resolve this by add `proxy_next_upstream error timeout invalid_header http_500 http_502 http_503 http_504 http_429 non_idempotent;` into APISIX config's `configurationSnippet.httpSrv`, and then set my backup server's priority to `-1`, every thing works fine~
> 
> But I have another problem now: APISIX's prometheus plguin seems do not have any metrics about retry, I can not monitor how frequently the retries happens through metrics. It's not a big problem, because the access logs can records all the `response_status` in order, I just curious is it possible to records the retries metrics in prometheus plugin.

That's a good user scenario. It's worth writing to the FAQ.

**Reply by ryan4yin** [reply]
and I customed the https://github.com/apache/apisix/blob/master/apisix/cli/ngx_tpl.lua to forward the `X-Forward-` prefixed headers to the mirror server.

I will create an issue latter to descibe this usage scenario and the problem I encountered, maybe this weekend.
